File: 3660. Jump Game IX/Solution.cpp
#include <climits>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> maxValue(vector<int>& nums)
    {
        vector<int> premax(nums.size());
        premax[0] = nums[0];
        for (size_t i = 1; i < nums.size(); ++i)
            premax[i] = max(premax[i - 1], nums[i]);
            
        int sufmin = INT_MAX;
        for (int i = nums.size() - 1; i >= 0; --i) {
            if (premax[i] > sufmin)
                premax[i] = premax[i + 1];
            sufmin = min(sufmin, nums[i]);
        }
        return premax;
    }
};

File: 3660. Jump Game IX/Solution.py
import math
from itertools import accumulate


class Solution:
    def maxValue(self, nums: list[int]) -> list[int]:
        n = len(nums)
        pre_max = list(accumulate(nums, max))

        suf_min = math.inf
        for i in range(n - 1, -1, -1):
            if pre_max[i] > suf_min:
                pre_max[i] = pre_max[i + 1]
            suf_min = min(suf_min, nums[i])
        return pre_max
